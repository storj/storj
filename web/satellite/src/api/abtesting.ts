// Copyright (C) 2022 Storj Labs, Inc.
// See LICENSE for copying information.

import { HttpClient } from '@/utils/httpClient';
import { ABHitAction, ABTestApi, ABTestValues } from '@/types/abtesting';

/**
 * ABHttpApi is a console AB testing API.
 * Exposes all ab-testing related functionality
 */
export class ABHttpApi implements ABTestApi {
    private readonly http: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/v0/ab';

    /**
     * Used to get test banner information.
     *
     * @throws Error
     */
    public async fetchABTestValues(): Promise<ABTestValues> {
        const path = `${this.ROOT_PATH}/values`;
        const response = await this.http.get(path);
        const abResponse = await response.json();

        if (!(response.ok && abResponse)) {
            return new ABTestValues();
        }

        return new ABTestValues(abResponse.has_new_banner);
    }

    public async sendHit(action: ABHitAction): Promise<void> {
        const path = `${this.ROOT_PATH}/hit/${action}`;
        await this.http.post(path, null);
    }
}
