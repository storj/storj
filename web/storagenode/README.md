# Storagenode web dashboard

## Development

The common commands to use when developing or making changes to this sources are mapped in the
package.json (`scripts` field).

To run any of them execute `npm run <command>`.

The most used commands for development are:
- `lint`: To lint the sources and fix issues that are auto-fixable.
- `dev`: To run the web server and watch changes to auto refresh.
- `test`: To run the unit tests.


### Load local changes in storj-up

Sometimes the unit tests and the local development server isn't enough to see how the new changes
look like with data and satellite API calls.

To visualize them, you can use [storj-up](https://github.com/storj/up).

storj-up uses docker compose to run Storj network on your local machine. It uses published docker
images to run, so it won't be see you local changes without indicating it.

To make storj-up to see your local changes in the Storagenode dashboard, you have to modify the
`docker-compose.yaml` file that it generates.

Before you must make a clean installation, and build the frontend with `npm ci & npm run build`.

If you made changes in the storanode service (backend), you must build the new binary for Linux
amd64. On Linux machine with the same architecture, you only need to execute from the root of this
repository `go build -o /some/path/storagenode ./cmd/storagenode`, on an Intel MacOS/OSX it is
something like
`CC=x86_64-unknown-linux-gnu-gcc GOOS=linux GOARCH=amd64 CGO_ENABLED=1 go build -o /some/path/storagenode ./cmd/storagenode`

Then, you must apply the following changes to one of the storagenode services changing the
paths accordingly to your local machine:
- Add a new volume that maps to the frontend (`source` is the path to your local machine to this
  directory):
  ```yaml
    volumes:
      - type: bind
        source: /some/path/storj/web/storagenode/
        target: /var/lib/storj/web/storagenode/
        bind:
          create_host_path: true
  ```
- Add a this new environment variable
  `STORJ_CONSOLE_STATIC_DIR: /var/lib/storj/web/storagenode`
- If you modified the storagenode backend, also add this volume (`source` is the path to your local
  machine where the new compiled binary is):
  ```yaml
    volumes:
      - type: bind
        source: /some/path/storagenode
        target: /var/lib/storj/go/bin/storagenode
        bind:
          create_host_path: true
  ```

Remember to run stop and run the docker services again.

You can find your modified frontend on http://localhost:3000
