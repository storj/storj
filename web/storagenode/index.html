<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <link rel="apple-touch-icon" href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAIAAAD8GO2jAAAC40lEQVRIDe1US0iUURQ+V8fJmcnykTJmNA0tMszETUVlCrkoR+2lWdFDk7AHlNQiCGaREhlUli3ETSRFqx4UlAi5ySgqS8hQiXLMTKca35k6j3O6d5zE//rnuNCF4Mdd3HvO+c53/3P/cxgRwUwiYCaTi9xzAn4rPPtLpPH7jaMBSDA4DE43AIFWAwYdBLApUf0L/OyFJ2/oWT28/wKd3Tw/RIdB4nLYnAjpa5gxzI8Mm6TRRlxQ8ZSuPQSbjYEGjEaICBHpegagww7gBpOJTm6HY+lMp/2/DBdQRfcA7bmIkEIaC+Vexqq3ZLPTkJOGndT6g6rfUf5V1FqIB+wqRkefag5hBFXP7yHKvoCQRHEFPBeOxgz8oZeNYvHNKGrqMeE48bAdRdj/zyglVBFApLM3EZJpXSHyy3K4PXTjMZpzMTCN+DIfwrJH6HILV7uDks4gbKLTFejxCIsEFYG6T6jPoKjd9MHmu3vpA4RUCrLQViumWb2VSeVv4/M2tuHivaSz0KtGn2W8hopAQZm4vrVSRDd9o1PlqM8Ukvdqffz7L3B+JkXm4OtmX6riu4KSd2UKAl39ZDpIoVnU4cBWO8XmixIv2EZHrivIWd4X0mfguVs44qLObozIwZj9ZO8Zf3uxl0dFUxt0dMHaFRQdwSprqPkjZG+h56VQkqfoqxPpULiPQgys5A6rriNjGNuwEjq7gNMlyAL2XnANQpxJhCXHs3IrXDrMEswQ7u2AMXLKalZ6lJ0/QIhQVSfMnIKDYO8ZC/Ft5E52ugA8EKIX7uR4sSbBrz4xOYK9XWaYJ4iCroQsYAgGpoPbNVD/WRk44TTsotoGFqyDneuFz9EPEAScLkEWWBoF4aHQ8p21tEuRE47ElkVTUS5sXMU8KCaVIRQ4XYI8i1xuaG4nMTX9IZBBzCIWuVDE8d5s+Ar8N4hdwrRBCqYsoHBOx0H+i6YjpyLHnICiHGqHuRKpVUVhm/0l+gsfrGM3Nzju5wAAAABJRU5ErkJggg==" type="image/x-icon">
        <meta name="apple-mobile-web-app-capable" content="yes">
        <meta name="apple-mobile-web-app-status-bar-style" content="black">
        <meta name="viewport" content="width=device-width, initial-scale=1, viewport-fit=cover">
        <meta name="description" content="Node Dashboard page">
        <link rel="shortcut icon" href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAIAAAD8GO2jAAAC40lEQVRIDe1US0iUURQ+V8fJmcnykTJmNA0tMszETUVlCrkoR+2lWdFDk7AHlNQiCGaREhlUli3ETSRFqx4UlAi5ySgqS8hQiXLMTKca35k6j3O6d5zE//rnuNCF4Mdd3HvO+c53/3P/cxgRwUwiYCaTi9xzAn4rPPtLpPH7jaMBSDA4DE43AIFWAwYdBLApUf0L/OyFJ2/oWT28/wKd3Tw/RIdB4nLYnAjpa5gxzI8Mm6TRRlxQ8ZSuPQSbjYEGjEaICBHpegagww7gBpOJTm6HY+lMp/2/DBdQRfcA7bmIkEIaC+Vexqq3ZLPTkJOGndT6g6rfUf5V1FqIB+wqRkefag5hBFXP7yHKvoCQRHEFPBeOxgz8oZeNYvHNKGrqMeE48bAdRdj/zyglVBFApLM3EZJpXSHyy3K4PXTjMZpzMTCN+DIfwrJH6HILV7uDks4gbKLTFejxCIsEFYG6T6jPoKjd9MHmu3vpA4RUCrLQViumWb2VSeVv4/M2tuHivaSz0KtGn2W8hopAQZm4vrVSRDd9o1PlqM8Ukvdqffz7L3B+JkXm4OtmX6riu4KSd2UKAl39ZDpIoVnU4cBWO8XmixIv2EZHrivIWd4X0mfguVs44qLObozIwZj9ZO8Zf3uxl0dFUxt0dMHaFRQdwSprqPkjZG+h56VQkqfoqxPpULiPQgys5A6rriNjGNuwEjq7gNMlyAL2XnANQpxJhCXHs3IrXDrMEswQ7u2AMXLKalZ6lJ0/QIhQVSfMnIKDYO8ZC/Ft5E52ugA8EKIX7uR4sSbBrz4xOYK9XWaYJ4iCroQsYAgGpoPbNVD/WRk44TTsotoGFqyDneuFz9EPEAScLkEWWBoF4aHQ8p21tEuRE47ElkVTUS5sXMU8KCaVIRQ4XYI8i1xuaG4nMTX9IZBBzCIWuVDE8d5s+Ar8N4hdwrRBCqYsoHBOx0H+i6YjpyLHnICiHGqHuRKpVUVhm/0l+gsfrGM3Nzju5wAAAABJRU5ErkJggg==" type="image/x-icon">
        <title>Node Dashboard</title>
    </head>
    <body>
        <div id="app"></div>
    </body>
</html>