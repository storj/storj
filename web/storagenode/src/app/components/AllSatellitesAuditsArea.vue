// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="audits-area">
        <div class="audits-area__content">
            <div v-for="item in auditItems" :key="item.satelliteName" class="audits-area__content__item">
                <div class="audits-area__content__item__header">
                    <p class="audits-area__content__item__header__satellite-name">{{ item.satelliteName }}</p>
                    <DisqualifyIcon v-if="item.iconClassName" :class="[ item.iconClassName ]" />
                </div>
                <div class="audits-area__content__item__divider" />
                <div class="audits-area__content__item__info">
                    <p class="audits-area__content__item__info__label">Suspension</p>
                    <p class="audits-area__content__item__info__value" :class="[ item.suspensionScore.statusClassName ]">{{ item.suspensionScore.label }}</p>
                </div>
                <div class="audits-area__content__item__info">
                    <p class="audits-area__content__item__info__label">Audit</p>
                    <p class="audits-area__content__item__info__value" :class="[ item.auditScore.statusClassName ]">{{ item.auditScore.label }}</p>
                </div>
                <div class="audits-area__content__item__info">
                    <p class="audits-area__content__item__info__label">Online</p>
                    <p class="audits-area__content__item__info__value" :class="[ item.onlineScore.statusClassName ]">{{ item.onlineScore.label }}</p>
                </div>
                <div class="audits-area__content__item__info">
                    <p class="audits-area__content__item__info__label">Vetted</p>
                    <p class="audits-area__content__item__info__value" :class="[ getVettedStatusClass(item.satelliteName) ]">{{ getVettedStatusLabel(item.satelliteName) }}</p>
                </div>
            </div>
        </div>
        <div v-if="isLoadMoreButtonVisible" class="audits-area__load-more-button" @click="loadMore">
            <p class="audits-area__load-more-button__label">Load More</p>
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import { SatelliteInfo, SatelliteScores } from '@/storagenode/sno/sno';

import DisqualifyIcon from '@/../static/images/disqualify.svg';

// @vue/component
@Component({
    components: { DisqualifyIcon },
})
export default class AllSatellitesAuditsArea extends Vue {
    /**
     * Number of score blocks displayed on page.
     */
    public numberOfItemsOnPage = 6;
    /**
     * Number of blocks added to displayed on page by clocking "Load more".
     */
    private readonly ITEMS_TO_ADD_COUNT: number = 6;

    /**
     * Returns reduced number of satellites score items depends on numberOfItemsOnPage.
     */
    public get auditItems(): SatelliteScores[] {
        return this.satellitesScores.slice(0, this.numberOfItemsOnPage);
    }

    /**
     * Indicates if all existing items are shown on page.
     */
    public get isLoadMoreButtonVisible(): boolean {
        return this.auditItems.length !== this.satellitesScores.length;
    }

    /**
     * Returns list of satellites score from store.
     */
    private get satellitesScores(): SatelliteScores[] {
        return this.$store.state.node.satellitesScores;
    }

    /**
     * Returns list of satellites from store.
     */
    private get satellites(): SatelliteInfo[] {
        return this.$store.state.node.satellites;
    }

    /**
     * Increments number of shown satellite score items by ITEMS_TO_ADD_COUNT.
     */
    public loadMore(): void {
        this.numberOfItemsOnPage += this.ITEMS_TO_ADD_COUNT;
    }

    /**
     * Returns vetted status label for satellite.
     */
    public getVettedStatusLabel(satelliteName: string): string {
        const satellite = this.findSatelliteByName(satelliteName);
        if (satellite?.vettedAt) {
            return satellite.vettedAt.toLocaleDateString();
        }
        return 'Not vetted';
    }

    /**
     * Returns vetted status CSS class for satellite.
     */
    public getVettedStatusClass(satelliteName: string): string {
        const satellite = this.findSatelliteByName(satelliteName);
        return satellite?.vettedAt ? 'vetted' : 'not-vetted';
    }

    /**
     * Finds satellite by name in satellites list.
     */
    private findSatelliteByName(satelliteName: string): SatelliteInfo | undefined {
        // SatelliteScores uses satelliteName but SatelliteInfo uses url
        // We need to match by URL since that's what's typically shown as the name
        return this.satellites.find(satellite => satellite.url === satelliteName);
    }
}
</script>

<style scoped lang="scss">
    .audits-area {
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;

        &__content {
            width: 100%;
            display: grid;
            grid-gap: 15px;
            grid-template-columns: repeat(3, 1fr);

            &__item {
                padding: 12px 16px;
                display: flex;
                flex-direction: column;
                align-items: flex-start;
                justify-content: center;
                color: var(--regular-text-color);
                background-color: var(--block-background-color);
                font-size: 14px;
                border-radius: 10px;

                &__header {
                    display: flex;
                    align-items: center;
                    justify-content: space-between;
                    width: 100%;

                    &__satellite-name {
                        font-family: 'font_medium', sans-serif;
                    }
                }

                &__divider {
                    width: 100%;
                    background: #e5e9f1;
                    height: 1px;
                    margin: 12px 0;
                }

                &__info {
                    display: flex;
                    align-items: center;
                    justify-content: space-between;
                    width: 100%;

                    &__label {
                        font-family: 'font_regular', sans-serif;
                    }

                    &__value {
                        font-family: 'font_bold', sans-serif;
                        font-size: 16px;
                    }
                }
            }
        }

        &__load-more-button {
            display: flex;
            align-items: center;
            justify-content: center;
            width: 226px;
            height: 48px;
            border: 1px solid #afb7c1;
            box-sizing: border-box;
            border-radius: 8px;
            cursor: pointer;
            margin-top: 25px;

            &__label {
                font-family: 'font_medium', sans-serif;
                font-size: 16px;
                color: var(--regular-text-color);
            }
        }
    }

    .disqualification {
        color: var(--critical-color);

        ::v-deep path {
            fill: var(--critical-color);
        }
    }

    .warning {
        color: var(--warning-color);

        ::v-deep path {
            fill: var(--warning-color);
        }
    }

    .vetted {
        color: var(--success-color, #00bf5f);
    }

    .not-vetted {
        color: var(--warning-color);
    }

    @media screen and (max-width: 800px) {

        .audits-area__content {
            grid-template-columns: repeat(2, 1fr);
        }
    }

    @media screen and (max-width: 500px) {

        .audits-area__content {
            grid-template-columns: 1fr;
        }
    }
</style>
