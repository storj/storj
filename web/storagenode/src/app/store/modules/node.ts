// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/app/store';
import { StorageNodeState } from '@/app/types/sno';
import { Duration, millisecondsInSecond, secondsInMinute } from '@/app/utils/duration';
import { getMonthsBeforeNow } from '@/app/utils/payout';
import { StorageNodeService } from '@/storagenode/sno/service';
import {
    Dashboard,
    Node,
    Satellite,
    SatelliteInfo,
    Satellites,
    Utilization,
} from '@/storagenode/sno/sno';

export const NODE_MUTATIONS = {
    POPULATE_STORE: 'POPULATE_STORE',
    SELECT_SATELLITE: 'SELECT_SATELLITE',
    SELECT_ALL_SATELLITES: 'SELECT_ALL_SATELLITES',
    SET_DAILY_DATA: 'SET_DAILY_DATA',
};

export const NODE_ACTIONS = {
    GET_NODE_INFO: 'GET_NODE_INFO',
    SELECT_SATELLITE: 'SELECT_SATELLITE',
};

export const StatusOnline = 'Online';
export const StatusOffline = 'Offline';

export const QUIC_STATUS = {
    StatusOk: 'OK',
    StatusMisconfigured: 'Misconfigured',
    StatusRefreshing: 'Refreshing',
};

const {
    POPULATE_STORE,
    SELECT_SATELLITE,
    SELECT_ALL_SATELLITES,
    SET_DAILY_DATA,
} = NODE_MUTATIONS;

const STATUS_TRESHHOLD_MINUTES = 120;

interface StorageNodeContext {
    state: StorageNodeState;
    commit: (string, ...unknown) => void;
}

export function newNodeModule(service: StorageNodeService): StoreModule<StorageNodeState> {
    return {
        state: new StorageNodeState(),
        mutations: {
            [POPULATE_STORE](state: StorageNodeState, nodeInfo: Dashboard): void {
                const minutesPassed = Duration.difference(new Date(), new Date(nodeInfo.lastPinged)) / millisecondsInSecond / secondsInMinute;
                const status = minutesPassed < STATUS_TRESHHOLD_MINUTES ? StatusOnline : StatusOffline;
                state.info = new Node(
                    nodeInfo.nodeID,
                    status,
                    nodeInfo.lastPinged,
                    nodeInfo.startedAt,
                    nodeInfo.version,
                    nodeInfo.allowedVersion,
                    nodeInfo.wallet,
                    nodeInfo.walletFeatures,
                    nodeInfo.isUpToDate,
                    nodeInfo.quicStatus,
                    nodeInfo.configuredPort,
                    nodeInfo.lastQuicPingedAt,
                );

                state.utilization = new Utilization(
                    nodeInfo.bandwidth,
                    nodeInfo.diskSpace,
                );

                state.disqualifiedSatellites = nodeInfo.satellites.filter((satellite: SatelliteInfo) => satellite.disqualified);
                state.suspendedSatellites = nodeInfo.satellites.filter((satellite: SatelliteInfo) => satellite.suspended);

                state.satellites = nodeInfo.satellites;
            },
            [SELECT_SATELLITE](state: StorageNodeState, satelliteInfo: Satellite): void {
                const selectedSatellite = state.satellites.find(satellite => satelliteInfo.id === satellite.id);

                if (!selectedSatellite) {
                    return;
                }

                state.selectedSatellite = new SatelliteInfo(
                    satelliteInfo.id,
                    selectedSatellite.url,
                    selectedSatellite.disqualified,
                    selectedSatellite.suspended,
                    selectedSatellite.vettedAt,
                    satelliteInfo.joinDate,
                );

                state.audits = satelliteInfo.audits;
            },
            [SELECT_ALL_SATELLITES](state: StorageNodeState, satelliteInfo: Satellites): void {
                state.selectedSatellite = new SatelliteInfo(
                    '',
                    '',
                    null,
                    null,
                    null,
                    satelliteInfo.joinDate,
                );
                state.satellitesScores = satelliteInfo.satellitesScores;
            },
            [SET_DAILY_DATA](state: StorageNodeState, satelliteInfo: Satellite): void {
                state.bandwidthChartData = satelliteInfo.bandwidthDaily;
                state.egressChartData = satelliteInfo.egressDaily;
                state.ingressChartData = satelliteInfo.ingressDaily;
                state.storageChartData = satelliteInfo.storageDaily;
                state.bandwidthSummary = satelliteInfo.bandwidthSummary;
                state.egressSummary = satelliteInfo.egressSummary;
                state.ingressSummary = satelliteInfo.ingressSummary;
                state.storageSummary = satelliteInfo.storageSummary;
                state.averageUsageBytes = satelliteInfo.averageUsageBytes;
            },
        },
        actions: {
            [NODE_ACTIONS.GET_NODE_INFO]: async function ({ commit }: StorageNodeContext): Promise<void> {
                const dashboard = await service.dashboard();

                commit(NODE_MUTATIONS.POPULATE_STORE, dashboard);
            },
            [NODE_ACTIONS.SELECT_SATELLITE]: async function ({ commit }: StorageNodeContext, id?: string): Promise<void> {
                let response: Satellite | Satellites;
                if (id) {
                    response = await service.satellite(id);
                    commit(NODE_MUTATIONS.SELECT_SATELLITE, response);
                } else {
                    response = await service.satellites();
                    commit(NODE_MUTATIONS.SELECT_ALL_SATELLITES, response);
                }

                commit(NODE_MUTATIONS.SET_DAILY_DATA, response);
            },
        },
        getters: {
            monthsOnNetwork: (state): number => {
                return getMonthsBeforeNow(state.selectedSatellite.joinDate);
            },
        },
    };
}
