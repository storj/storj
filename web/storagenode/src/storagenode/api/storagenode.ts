// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    Dashboard,
    Satellite,
    SatelliteByDayInfo,
    SatelliteInfo,
    Satellites,
    SatelliteScores,
    Traffic,
} from '@/storagenode/sno/sno';
import { HttpClient } from '@/storagenode/utils/httpClient';

/**
 * Used to get dashboard and satellite data from json.
 */
export class StorageNodeApi {
    private readonly client: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/sno';

    /**
     * Gets dashboard data from server.
     * @returns dashboard - new dashboard instance filled with data from json.
     */
    public async dashboard(): Promise<Dashboard> {
        const response = await this.client.get(this.ROOT_PATH);

        if (!response.ok) {
            throw new Error('can not get node information');
        }

        const data = await response.json();

        const satellitesJson = data.satellites || [];

        const satellites: SatelliteInfo[] = satellitesJson.map((satellite: any) => { // eslint-disable-line @typescript-eslint/no-explicit-any
            const disqualified: Date | null = satellite.disqualified ? new Date(satellite.disqualified) : null;
            const suspended: Date | null = satellite.suspended ? new Date(satellite.suspended) : null;
            const vettedAt: Date | null = satellite.vettedAt ? new Date(satellite.vettedAt) : null;

            return new SatelliteInfo(satellite.id, satellite.url, disqualified, suspended, vettedAt);
        });

        const diskSpace: Traffic = new Traffic(data.diskSpace.used, data.diskSpace.available, data.diskSpace.trash, data.diskSpace.overused);
        const bandwidth: Traffic = new Traffic(data.bandwidth.used);

        return new Dashboard(data.nodeID, data.wallet, data.walletFeatures || [], satellites, diskSpace, bandwidth,
            new Date(data.lastPinged), new Date(data.startedAt), data.version, data.allowedVersion, data.upToDate, data.quicStatus, data.configuredPort, new Date(data.lastQuicPingedAt));
    }

    /**
     * Gets satellite data from server.
     * @returns satellite - new satellite instance filled with data from json.
     */
    public async satellite(id: string): Promise<Satellite> {
        const url = `${this.ROOT_PATH}/satellite/${id}`;

        const response = await this.client.get(url);

        if (!response.ok) {
            throw new Error('can not get satellite information');
        }

        const data = await response.json();

        const satelliteByDayInfo = new SatelliteByDayInfo(data);

        const audits: SatelliteScores = new SatelliteScores(
            data.audits.satelliteName,
            data.audits.auditScore,
            data.audits.suspensionScore,
            data.audits.onlineScore,
        );

        return new Satellite(
            data.id,
            satelliteByDayInfo.storageDaily,
            satelliteByDayInfo.bandwidthDaily,
            satelliteByDayInfo.egressDaily,
            satelliteByDayInfo.ingressDaily,
            data.storageSummary,
            data.averageUsageBytes,
            data.bandwidthSummary,
            data.egressSummary,
            data.ingressSummary,
            audits,
            new Date(data.nodeJoinedAt),
        );
    }

    /**
     * Gets data for all satellites from server.
     * @returns satellites - new satellites instance filled with data from json.
     */
    public async satellites(): Promise<Satellites> {
        const url = `${this.ROOT_PATH}/satellites`;

        const response = await this.client.get(url);

        if (!response.ok) {
            throw new Error('can not get all satellites information');
        }

        const data = await response.json();

        const satelliteByDayInfo = new SatelliteByDayInfo(data);

        const satellitesScores = data.audits.map(scoreInfo => {
            return new SatelliteScores(
                scoreInfo.satelliteName,
                scoreInfo.auditScore,
                scoreInfo.suspensionScore,
                scoreInfo.onlineScore,
            );
        });

        return new Satellites(
            satelliteByDayInfo.storageDaily,
            satelliteByDayInfo.bandwidthDaily,
            satelliteByDayInfo.egressDaily,
            satelliteByDayInfo.ingressDaily,
            data.storageSummary,
            data.averageUsageBytes,
            data.bandwidthSummary,
            data.egressSummary,
            data.ingressSummary,
            new Date(data.earliestJoinedAt),
            satellitesScores,
        );
    }
}
