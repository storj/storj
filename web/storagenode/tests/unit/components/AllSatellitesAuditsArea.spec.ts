// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';
import { createLocalVue, shallowMount } from '@vue/test-utils';

import { newNodeModule, NODE_ACTIONS, NODE_MUTATIONS } from '@/app/store/modules/node';
import { StorageNodeApi } from '@/storagenode/api/storagenode';
import { StorageNodeService } from '@/storagenode/sno/service';
import { Dashboard, SatelliteInfo, Satellites, SatelliteScores, Traffic } from '@/storagenode/sno/sno';

import AllSatellitesAuditsArea from '@/app/components/AllSatellitesAuditsArea.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

const nodeApi = new StorageNodeApi();
const nodeService = new StorageNodeService(nodeApi);
const nodeModule = newNodeModule(nodeService);

const store = new Vuex.Store({ modules: { node: nodeModule } });

describe('AllSatellitesAuditsArea', (): void => {

    it('renders correctly with actual values', async (): Promise<void> => {
        const vettedDate = new Date(2023, 0, 15);

        // Mock dashboard data with satellites
        const dashboardInfo = new Dashboard(
            '1',
            '2',
            [],
            [
                new SatelliteInfo('sat1', 'name1', null, null, vettedDate),
                new SatelliteInfo('sat2', 'name2', null, null, null),
                new SatelliteInfo('sat3', 'name3', null, null, vettedDate),
            ],
            new Traffic(99, 100, 4),
            new Traffic(50),
            new Date(),
            new Date(),
            '0.1.1',
            '0.2.2',
            false,
            'OK',
            '13000',
            new Date(),
        );

        // Populate store with satellite info
        store.commit(NODE_MUTATIONS.POPULATE_STORE, dashboardInfo);

        const satellites = new Satellites();
        satellites.satellitesScores = [
            new SatelliteScores('name1', 1, 1, 0.5),
            new SatelliteScores('name2', 0.5, 1, 0.97),
            new SatelliteScores('name3', 0.97, 1, 1),
        ];

        jest.spyOn(nodeApi, 'satellites').mockReturnValue(Promise.resolve(satellites));

        await store.dispatch(NODE_ACTIONS.SELECT_SATELLITE);

        const wrapper = shallowMount(AllSatellitesAuditsArea, {
            store,
            localVue,
        });

        expect(wrapper).toMatchSnapshot();
    });
});
