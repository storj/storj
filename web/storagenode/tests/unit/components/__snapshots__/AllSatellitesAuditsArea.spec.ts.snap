// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllSatellitesAuditsArea renders correctly with actual values 1`] = `
<div class="audits-area">
  <div class="audits-area__content">
    <div class="audits-area__content__item">
      <div class="audits-area__content__item__header">
        <p class="audits-area__content__item__header__satellite-name">name1</p>
        <disqualifyicon-stub class="disqualification"></disqualifyicon-stub>
      </div>
      <div class="audits-area__content__item__divider"></div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Suspension</p>
        <p class="audits-area__content__item__info__value">100 %</p>
      </div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Audit</p>
        <p class="audits-area__content__item__info__value">100 %</p>
      </div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Online</p>
        <p class="audits-area__content__item__info__value disqualification">50 %</p>
      </div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Vetted</p>
        <p class="audits-area__content__item__info__value vetted">1/15/2023</p>
      </div>
    </div>
    <div class="audits-area__content__item">
      <div class="audits-area__content__item__header">
        <p class="audits-area__content__item__header__satellite-name">name2</p>
        <disqualifyicon-stub class="disqualification"></disqualifyicon-stub>
      </div>
      <div class="audits-area__content__item__divider"></div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Suspension</p>
        <p class="audits-area__content__item__info__value">100 %</p>
      </div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Audit</p>
        <p class="audits-area__content__item__info__value disqualification">50 %</p>
      </div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Online</p>
        <p class="audits-area__content__item__info__value warning">97 %</p>
      </div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Vetted</p>
        <p class="audits-area__content__item__info__value not-vetted">Not vetted</p>
      </div>
    </div>
    <div class="audits-area__content__item">
      <div class="audits-area__content__item__header">
        <p class="audits-area__content__item__header__satellite-name">name3</p>
        <disqualifyicon-stub class="warning"></disqualifyicon-stub>
      </div>
      <div class="audits-area__content__item__divider"></div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Suspension</p>
        <p class="audits-area__content__item__info__value">100 %</p>
      </div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Audit</p>
        <p class="audits-area__content__item__info__value warning">97 %</p>
      </div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Online</p>
        <p class="audits-area__content__item__info__value">100 %</p>
      </div>
      <div class="audits-area__content__item__info">
        <p class="audits-area__content__item__info__label">Vetted</p>
        <p class="audits-area__content__item__info__value vetted">1/15/2023</p>
      </div>
    </div>
  </div>
  <!---->
</div>
`;
